File: frontend/src/lib/api/recurring.ts
import apiClient from './client'

export interface RecurringDefinition {
  period: string
  account: string
  counter_account: string
  amount: number
  currency: string
  payee?: string | null
}

export interface CreateRecurringRequest {
  period: string
  account: string
  counter_account: string
  amount: string
  currency?: string
  payee?: string
}

export interface RecurringDefinitionResponse {
  formatted_text: string
  definition: RecurringDefinition
}

export interface RecurringForecastResponse {
  output: string
}

export const recurringApi = {
  list: (workspaceId: string) =>
    apiClient.get<RecurringDefinition[]>(`/workspaces/${workspaceId}/recurring`),

  create: (workspaceId: string, data: CreateRecurringRequest) =>
    apiClient.post<RecurringDefinitionResponse>(`/workspaces/${workspaceId}/recurring`, data),

  update: (workspaceId: string, index: number, data: CreateRecurringRequest) =>
    apiClient.put<RecurringDefinitionResponse>(
      `/workspaces/${workspaceId}/recurring/${index}`,
      data,
    ),

  delete: (workspaceId: string, index: number) =>
    apiClient.delete(`/workspaces/${workspaceId}/recurring/${index}`),

  forecast: (workspaceId: string, endDate?: string) =>
    apiClient.get<RecurringForecastResponse>(`/workspaces/${workspaceId}/recurring/forecast`, {
      params: endDate ? { end_date: endDate } : undefined,
    }),
}

File: backend/src/models/v1/recurring.rs
use serde::{Deserialize, Serialize};

/// A single recurring transaction — persisted as a ledger periodic-transaction
/// (`~ <period>`) block in the workspace's `recurring.ledger` file.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct RecurringDefinition {
    pub period: String,
    /// Destination account, e.g. `Expenses:Subscriptions:Netflix`.
    pub account: String,
    /// The paying-side account, e.g. `Assets:Bank:Revolut`.
    pub counter_account: String,
    pub amount: f64,
    pub currency: String,
    /// Optional payee label captured as a comment next to the posting.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub payee: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct CreateRecurringRequest {
    pub period: String,
    pub account: String,
    pub counter_account: String,
    pub amount: String,
    pub currency: Option<String>,
    pub payee: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct UpdateRecurringRequest {
    pub period: String,
    pub account: String,
    pub counter_account: String,
    pub amount: String,
    pub currency: Option<String>,
    pub payee: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct RecurringForecastQuery {
    pub end_date: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct RecurringDefinitionResponse {
    pub formatted_text: String,
    pub definition: RecurringDefinition,
}

#[derive(Debug, Clone, Serialize)]
pub struct RecurringForecastResponse {
    /// Raw `ledger register --forecast` output; frontend parses same as other register queries.
    pub output: String,
}
